### PR TITLE
HTTPS Upgrades Test Page

### PR DESCRIPTION
**Pull request by kdzwinel**
Tech Design: https://app.asana.com/0/0/1198919147577402/f
Preview: http://privacy-test-pages.glitch.me/privacy-protections/https-upgrades/

Please note that preview page returns HTTP instead of HTTPS for now. This is until our extension will actually start upgrading good.trird-party.site to HTTPS.

As for the websocket - it fails, but this is a known glitch limitation that they are fixing.

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -40,10 +40,11 @@ <h2>Security</h2>
...
 <ul>
-	<li><a href="./privacy-protections/request-blocking/">Request blocking</a></li>
-	<li><a href='./privacy-protections/fingerprinting/'>Fingerprinting</a></li>
-	<li><a href='./privacy-protections/storage-blocking/'>Storage blocking</a></li>
-	<li><a href='./privacy-protections/referrer-trimming/'>Referrer trimming</a></li>
+	<li><a href='https://privacy-test-pages.glitch.me/privacy-protections/request-blocking/'>Request blocking</a></li>
+	<li><a href='https://privacy-test-pages.glitch.me/privacy-protections/fingerprinting/'>Fingerprinting</a></li>
+	<li><a href='https://privacy-test-pages.glitch.me/privacy-protections/storage-blocking/'>Storage blocking</a></li>
+	<li><a href='https://privacy-test-pages.glitch.me/privacy-protections/referrer-trimming/'>Referrer trimming</a></li>
+	<li><a href='http://privacy-test-pages.glitch.me/privacy-protections/https-upgrades/'>HTTPS upgrades</a></li>
```
Comment (kdzwinel) [inline index.html:47]: Some of those pages don't really work on github pages, which might be confusing, so I hardcoded full glitch urls.

File: privacy-protections/fingerprinting/index.html [hunk 2]
```diff
@@ -12,14 +12,15 @@
...
 
     <p>This page collects various datapoints from your browser that can be used for fingerprinting.</p>
     <p>⚠️ Please note that: </p>
     <ul>
         <li>not all of the properties listed here have to be modified to succesfully prevent fingerprining,</li>
-        <li>some properties may change randomly, depending on the brower, on every page reload or test run,</li>
-        <li>some data is duplicated, but it's retrieved using different methods.</li>
+        <li>depending on the brower some properties may change randomly on every page reload or test run,</li>
+        <li>some data is duplicated, but it's retrieved using different methods,</li>
+        <li>this page has to be loaded over https for all tests to work correctly (some APIs are not available over http).</li>
```
Comment (kdzwinel) [inline privacy-protections/fingerprinting/index.html:23]: Since glitch is supporting both http and https I added a note that fingerprinting page should be loaded over https.

File: privacy-protections/https-upgrades/frame.html [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    <script>
+        const onMessage = msg => {
+            if (msg.data.action && msg.data.action === 'url') {
+                if (window.opener) {
+                    window.opener.postMessage({url: document.location.href, type: msg.data.type}, '*');
+                } else if (window.parent) {
+                    window.parent.postMessage({url: document.location.href, type: msg.data.type}, '*');
+                }
+            }
+        };
```
Comment (kdzwinel) [inline privacy-protections/https-upgrades/frame.html:18]: this page returns current url to either parent url (iframe scenario) or opener (window.open scenario).

File: privacy-protections/index.html [hunk 4]
```diff
@@ -6,15 +6,16 @@
...
     <ul>
-        <li><a href='./request-blocking/'>Request blocking</a></li>
-        <li><a href='./fingerprinting/'>Fingerprinting</a></li>
-        <li><a href='./storage-blocking/'>Storage blocking</a></li>
-        <li><a href='./referrer-trimming/'>Referrer trimming</a></li>
+        <li><a href='https://privacy-test-pages.glitch.me/privacy-protections/request-blocking/'>Request blocking</a></li>
+        <li><a href='https://privacy-test-pages.glitch.me/privacy-protections/fingerprinting/'>Fingerprinting</a></li>
+        <li><a href='https://privacy-test-pages.glitch.me/privacy-protections/storage-blocking/'>Storage blocking</a></li>
+        <li><a href='https://privacy-test-pages.glitch.me/privacy-protections/referrer-trimming/'>Referrer trimming</a></li>
+        <li><a href='http://privacy-test-pages.glitch.me/privacy-protections/https-upgrades/'>HTTPS upgrades</a></li>
```
Comment (kdzwinel) [inline privacy-protections/index.html:18]: ditto - hardcoded urls

File: privacy-protections/referrer-trimming/main.js [hunk 5]
```diff
@@ -22,7 +22,7 @@ function generateNavigationTest(url) {
 
     if (localStorage[key]) {// test already finished before
         return JSON.parse(localStorage[key]);
-    } else if(currentURL.searchParams.get('js')) {// test finished now
+    } else if(currentURL.searchParams.get('js') !== null) {// test finished now
```
Comment (kdzwinel) [inline privacy-protections/referrer-trimming/main.js:25]: unrelated fix for referrer trimming - we have to test for null specifically because Brave, in certain scenarios trims referrer to an empty string.

File: server.js [hunk 6]
```diff
@@ -2,6 +2,15 @@ const express = require('express');
...
 const app = express();
 const port = process.env.PORT || 3000;
+const url = require('url');
+
+function fullUrl(req) {
+  return url.format({
+    protocol: req.protocol,
+    host: req.get('host'),
+    pathname: req.originalUrl
+  });
```
Comment (kdzwinel) [inline server.js:12]: this returns full url that was requested (we need that to see if the protocol was http or https)

File: server.js [hunk 7]
```diff
@@ -108,9 +117,9 @@ app.get('/come-back', (req, res) => {
 <body>
 <script>
     const jsReferrer = document.referrer;
-    document.body.innerHTML += '<p>header: <strong>${req.headers.referer}</strong></p><p>js: <strong>' + jsReferrer + '</strong></p>';
+    document.body.innerHTML += '<p>header: <strong>${req.headers.referer || ''}</strong></p><p>js: <strong>' + jsReferrer + '</strong></p>';
```
Comment (kdzwinel) [inline server.js:120]: unrelated referrer clean up - for Brave referrer is sometimes undefined in which scenarios we want to return `''` instead `'undefined'`

**Review by kdzwinel — COMMENTED** [review]
Hey @Charlie-belmer PTAL whenever you'll have a moment. There are few unrelated changes bundled here sorry for that.

**Review by Charlie-belmer — APPROVED** [review]
Everything looks great to me Konrad!
